### PR TITLE
kubetest kops should wait for Ready nodes

### PR DESCRIPTION
**Pull request by justinsb**
At least for kops, we don't consider a cluster to be ready until the
nodes are actually Ready.

### CODE REVIEW

**Comment by justinsb** [comment]
cc @krzyzacy

**Comment by krzyzacy** [comment]
maybe we should also move `waitForNodes()` here?

**Comment by justinsb** [comment]
> maybe we should also move waitForNodes() here?

Yes, but I figured we should do so gradually (i.e. find any problems on the kops jobs before letting it loose on the other jobs).

This might also be a way to break up the behemoth that #4434 has become now, in particular looking at introducing context.Context if we decide we want it cc @fejta

**Comment by krzyzacy** [comment]
qq: we got a few [1.7 kops-aws on prow](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/52755/pull-kubernetes-e2e-kops-aws-prow/17/build-log.txt) seems having the same node not found issue, similar with the kops-gce one today. Probably need to resolve that before we restrict the condition further.

File: kubetest/kops.go [hunk 1]
```diff
@@ -188,7 +188,7 @@ func (k kops) Up() error {
 	// TODO(zmerlynn): More cluster validation. This should perhaps be
 	// added to kops and not here, but this is a fine place to loop
 	// for now.
-	return waitForNodes(k, k.nodes+1, *kopsUpTimeout)
+	return waitForReadyNodes(k.nodes+1, *kopsUpTimeout)
```
Comment (fejta) [inline kubetest/kops.go]: Any reason we want to use waitForNodes with kubernetes-anywhere? Can we just delete the previous logic?

Comment (justinsb) [reply kubetest/kops.go]: I do think we should change waitForNodes -> waitForReadyNodes everywhere, I just wanted to start with kops.  I can add a `TODO(justinsb)` here.

Comment (fejta) [reply kubetest/kops.go]: I'd vote for just making the change everywhere now. We don't need backwards compat for ci :)

IMO it is more important to kill dead code

**Comment by justinsb** [comment]
OK I can do that - there's also some additional work in #4434 though, and I'm happy to rebase that on top of this if we want to merge this first.  #4434 was originally just about capturing logs, but then there were some additional improvements requested - to implement a timeout (which I did via a Context object), and to add unit test (which I did, but made that PR even bigger).

My suggestion - we merge this one as it'll make #4434 a little smaller!  I'll move everything to use `waitForReadyNodes` though, and I'll comment on #4434 as well.

**Comment by justinsb** [comment]
Ah - and @krzyzacy the link with the 1.7 failure is that some of the logs are missing in both cases, and it would be great to have log-capture that didn't depend on something shipped in the k8s tree (because then we have to remain compatible with old versions, and we can't add capture of more logs easily if we discover another log would be useful).

And thanks for LGTM @fejta - will do waitForReadyNodes everywhere

**Comment by justinsb** [comment]
Updated so kube-anywhere also calls `waitForReadyNodes`, and removed `waitForNodes`

**Comment by krzyzacy** [comment]
do you feel like to add some unit tests to kubetest, start with the new codes?

**Comment by justinsb** [comment]
So yes, @krzyzacy I can certainly add unit tests - I actually added them in #4434.  I think I've been ill-disciplined in that I should fix the earliest PRs, even if the issue was commented on in a later PR in the chain.  The perils of trying to split PRs into more reviewable sequences!

I'll pull in the unit tests out of #4434 and that should make both more reviewable!

**Comment by justinsb** [comment]
Now with added unit tests :-)

File: kubetest/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	"encoding/json"
+	"fmt"
+	"log"
+	"os/exec"
+	"time"
+)
+
+// kubectlGetNodes lists nodes by executing kubectl get nodes, parsing the output into a nodeList object
+func kubectlGetNodes() (*nodeList, error) {
+	o, err := output(exec.Command("kubectl", "get", "nodes", "-ojson"))
```
Comment (krzyzacy) [inline kubetest/kubernetes.go]: just a question: do you care which kubectl you use here? (from gcloud or from k8s tarball)

**Comment by krzyzacy** [comment]
(that lint is fine)
/hold cancel
I'll bump another image

**Pull request merged**